00xx-account-deletion.md

```diff
@@ -0,0 +1,576 @@
+
+There is a single caveat: the comparison function for locations needs to be
+*total* for this choice to work. The implementations of `Location` in the
+codebase all have this property.
+
+#### On-disk ledger
+
+The
+[`Database`](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_ledger/database.ml)
+module implements an on-disk ledger that backs up the in-memory data structure.
+
+Deletion support relies on the same techniques as in-memory ledgers, that is, a
+representation of a ledger as Merkle tree *and* an additional free list.
+
+The main culprit is the representation of the free list on-disk. Here we will simply store
+the data on the KV database as follows:
+- the free list is identified by the key `free_list` (like
+  [`last_account_location`](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_ledger/database.ml#L235));
+- store the serialized encoding of the heap in the database.
+
+**Space requirements**. We argue that this is okay to have this simple scheme in
+terms of space use.  Indeed, in the worst case today, the ledger is full and we
+do need to store it in the database.  The requirement of storing the free list
+alongside the ledger have the same worst-case scenario. When deleting an account
+one is trading the storage of an account for the storage of a location, which is
+usually smaller (a single integer instead of a complex data structure).
+Therefore storing free locations does not add more storage requirements than
+today.
+
+The serialization of and array-based heap (ike `bheap`) of integers should be rather efficient.
+
+**Allocation cost**. Allocating locations from the free list now means
+1. read `free_list`
+2. deserialize the heap value;
+3. get the needed free locations;
+4. write the accounts to the free locations;
+5. reserialize the new value of the heap.
+
+**Simple improvement on batch allocations**. Furthermore, we will implement a
+decoding function where one can specify how many locations we ideally want to
+retrieve (up to `n`) to handle functions `set_location_batch`.  This function
+will return a list of free locations together with its size (so that we do not
+need to call `List.length`) in order to know if we need further allocation
+starting from the fill frontier index.  This function avoids
+deserializing/serializing multiple times when we know in advance we will do
+multiple allocations.
+
+There are some further details to be taken care of, such as:
+
+- patch function [`allocate`](https://github.com/MinaProtocol/mina/blob/4495af5caea5e1bb2f98f92592c065f93a586ade/src/lib/merkle_ledger/database.ml#L267) to handle the free list as well
+- updating Merkle paths on removal
+- updating the `all_accounts` function to avoid iterating over addresses that
+  have been removed
+
+Other strategies are discussed [here](#free_list_db alt).
+
+
+#### Syncing ledgers
```

Comment:
As well as synchronising ledgers, we also need to consider node startup for genesis. Consider the `ledger` field in the runtime configuration: we need a way to represent that an account is empty in that format, and faithfully convert it, so that the genesis ledger after a hard-fork faithfully represents the state of the ledger once we have account deletion.